Feature/RR-82 review goal

mattwills-moj-digital opened the pull request:
This PR introduces the review screen as per this screenshot:

![image](https://github.com/ministryofjustice/hmpps-education-and-work-plan-ui/assets/135589132/d57d1ed5-cea5-47c3-af00-a8e2c1a06ab9)

Note that this does not match the exact style in the prototype (which appears to use a blend of a GDS table and a summary list?), but the key functionality is there.

There are some question marks over how we should use an enum for the target date ranges (e.g. in `AddStepDto`), but I think we should address this as part of a separate PR.

nathanrussell-moj-digital commented on integration_tests/e2e/goal/addStep.cy.ts:
```diff
@@ -128,6 +129,6 @@ context('Add a step', () => {
 
     // Then
-    // assert we are on the next page
+    Page.verifyOnPage(AddNotePage)
```
Do we need to enable this test? (Line 111 ?)
Or; does the test not really add anything over and above the test in `addNote` above?, which is the 'screen to screen journey' ?

nathanrussell-moj-digital commented on integration_tests/e2e/goal/review.cy.ts:
```diff
@@ -0,0 +1,119 @@
+    cy.signIn()
+
+    // When
+    cy.visit(`/plan/${prisonNumber}/goals/add-note`)
```
This is navigating to add note, not review goal ?
(I think the rest of the test is sound; its just the `cy.visit` that is wrong?

nathanrussell-moj-digital commented on integration_tests/e2e/goal/review.cy.ts:
```diff
@@ -0,0 +1,119 @@
+    addNotePage //
+      .isForPrisoner(prisonNumber)
+      .submitPage()
+
```
Is it worth a quick assertion here to ensure we are on the Review Page at this point? (I know, assertions in the `// Given' block! 😢 )
```
const reviewPage = Page.verifyOnPage(ReviewPage)
reviewPage //
  .isForPrisoner(prisonNumber)
```

nathanrussell-moj-digital commented on integration_tests/e2e/goal/review.cy.ts:
```diff
@@ -0,0 +1,119 @@
+    overviewPage.isForPrisoner(someOtherPrisonNumber)
+  })
+
+  it('should not be able to navigate directly to review goal given Create Goal and Add Steps have been submitted but Add Note has not', () => {
```
Love it 👍 ❤️

nathanrussell-moj-digital commented on integration_tests/e2e/goal/review.cy.ts:
```diff
@@ -0,0 +1,119 @@
+    Page.verifyOnPage(AddNotePage)
+  })
+
+  it.skip('should move to overview page', () => {
```
This test looks like it could be a good one - any reason why its skipped?
It might be failing because you are trying to visit the create goal page directly without going via the Overview page? I think your first few lines of the test want to be something like this perhaps?
```
    // Given
    const prisonNumber = 'G6115VJ'
    cy.signIn()
    cy.visit(`/plan/${prisonNumber}/view/overview`)
    const overviewPage = Page.verifyOnPage(OverviewPage)

    const createGoalPage = overviewPage.clickAddGoalButton()
```

mattwills-moj-digital replied:
Fixed and enabled 👍

nathanrussell-moj-digital commented on server/routes/createGoal/routerRequestHandlers.test.ts:
```diff
@@ -211,4 +212,24 @@ describe('routerRequestHandlers', () => {
   })
+
+  describe('checkAddNoteFormExistsInSession', () => {
+    it(`should redirect to Create Goal screen given add note form does not exist in session`, async () => {
```
`should redirect to Add Note screen given add note form does not exist in session` ?

nathanrussell-moj-digital commented on server/routes/createGoal/routerRequestHandlers.ts:
```diff
@@ -45,6 +45,21 @@ const checkAddStepFormsArrayExistsInSession = async (req: Request, res: Response
+const checkAddNoteFormExistsInSession = async (req: Request, res: Response, next: NextFunction) => {
+  if (!req.session.addNoteForm) {
+    logger.warn(
+      `No AddNoteForm object in session - user attempting to navigate to path ${req.path} out of sequence. Redirecting to start of Create Goal journey.`,
```
`Redirecting to start of Create Goal journey.`
perhaps
`Redirecting to Add Note screen.` ?

nathanrussell-moj-digital commented on server/views/pages/goal/review/index.njk:
```diff
@@ -0,0 +1,61 @@
+
+{% block content %}
+
+  {% if errors.length > 0 %}
```
I think you can probably remove this if block, and the `error-summary` import, because there will never be any errors on this screen? (because the `ReviewView` object does not contain an `errors` property)

nathanrussell-moj-digital reviewed (commented):
Looks good. I've added a few "tidying" comments that might be worth looking at perhaps?
What do you think?

nathanrussell-moj-digital commented:
Also - re: your note in the description about it not matching the styles from the story screenshot / prototype ... What are your thoughts/plans re addressing this? Are you talking to Karen about it? Are we getting her to re-do her designs / update or comment on our story accepting the change?

Or (perhaps a loaded question 😁 ) ...  are you going to merge this is as to get the core functionality in place, then look at seeing how we can meet her visual designs a different way perhaps? I think we're in agreement that a summary card with a nested table is probably wrong .... are we going to look to see if we can write some css to style the table in the way the designs require? 😁 👍

chrimesdev commented:
> Also - re: your note in the description about it not matching the styles from the story screenshot / prototype ... What are your thoughts/plans re addressing this? Are you talking to Karen about it? Are we getting her to re-do her designs / update or comment on our story accepting the change?
> 
> Or (perhaps a loaded question 😁 ) ... are you going to merge this is as to get the core functionality in place, then look at seeing how we can meet her visual designs a different way perhaps? I think we're in agreement that a summary card with a nested table is probably wrong .... are we going to look to see if we can write some css to style the table in the way the designs require? 😁 👍

As we mentioned on Slack I think we should just ensure the functionality is working and I can retro fit the designs next week once I've managed to catchup with Karen! I think there's a hybrid of components being used and not sure its the best approach anyway, so might not be worth the effort if its likely to change.

mattwills-moj-digital commented:
> Looks good. I've added a few "tidying" comments that might be worth looking at perhaps? What do you think?

I won't lie, I was disappointed by my shoddiness with the integration tests! I should have looked at the diff properly before sharing this PR, but I've addressed the issues you spotted 👍 

I think we've covered your other question - let's get the functionality working first and then we (well Adam!) can look at the styles next week.

nathanrussell-moj-digital reviewed (approved):
LGTM; and thanks for making the proposed changes 👍
